controllers/authorController.js

```diff
@@ -15,35 +19,196 @@ exports.author_list = function(req, res) {
-exports.author_delete_post = function(req, res) {
-    res.send('NOT IMPLEMENTED: Author delete POST');
+exports.author_delete_post = function(req, res, next) {
+  async.parallel({
+      author: function(callback) {
+        Author.findById(req.body.authorid).exec(callback)
+      },
+      authors_blogposts: function(callback) {
+        BlogPost.find({ 'author': req.body.authorid }).exec(callback)
+      },
+  }, function(err, results) {
+      if (err) { return next(err); }
+      
+      // Success
+      if (results.authors_blogposts.length > 0) {
+          // Author has books. Render in same way as for GET route.
+          res.render('author_delete', { title: 'Delete Author', author: results.author, author_blogposts: results.authors_blogposts } );
+          return;
+      }
+      else {
+          // Author has no books. Delete object and redirect to the list of authors.
+          Author.findByIdAndRemove(req.body.authorid, function deleteAuthor(err) {
+              if (err) { return next(err); }
+              // Success - go to author list
+              res.redirect('/blog/authors')
+          })
+      }
+  });
 };
 
 // Display Author update form on GET.
-exports.author_update_get = function(req, res) {
-    res.send('NOT IMPLEMENTED: Author update GET');
+exports.author_update_get = function(req, res, next) {
+
+  async.parallel({
+      author: function(callback) {
+          Author.findById(req.params.id)
+          .exec(callback);
+      },
+      }, function(err, results) {
+          if (err) { return next(err); }
+          if (results.author==null) { // No results.
+              let err = new Error('Author not found');
+              err.status = 404;
+              return next(err);
+          }
+          // Success.
+          res.render('author_form', { title: 'Edit Author', author: results.author });
+          console.log(results.author)
+      });
 };
 
 // Handle Author update on POST.
-exports.author_update_post = function(req, res) {
-    res.send('NOT IMPLEMENTED: Author update POST');
-};
\ No newline at end of file
+exports.author_update_post = [
+
+  //Validate and Sanitise Fields
+  body('first_name').trim().isLength({ min: 1 }).escape().withMessage('First name must be specified.')
+      .isAlphanumeric().withMessage('First name has non-alphanumeric characters.'),
+  body('family_name').trim().isLength({ min: 1 }).escape().withMessage('Family name must be specified.')
+      .isAlphanumeric().withMessage('Family name has non-alphanumeric characters.'),
+  body('date_of_birth', 'Invalid date of birth').optional({ checkFalsy: true }).isISO8601().toDate(),
+  body('date_of_death', 'Invalid date of death').optional({ checkFalsy: true }).isISO8601().toDate(),
+
+    //Process request
+    (req, res, next) => {
+
+      //Extract validation errors
+      const errors = validationResult(req);
+
+      //Create an author object with escaped trimmed data and old ID.
+      let author = new Author(
```

Comment:
Why are you creating a new object here? Especially as it’s not actually being used?

Reply:
I thought that it got used below, either in the if statement to re-render the form with the new info the user has input if there are errors:

`res.render('author_form', { title: 'Edit Author', author: author, errors: errors.array() })`

, or in the else statement to update the existing Author document:

```
Author.findByIdAndUpdate(req.params.id, author, {}, function (err, theauthor) {
            if (err) { return next(err); }
```

Is there a better way to do this?